omisego-sdk/src/main/java/co/omisego/omisego/model/transaction/TransactionListParams.kt

```diff
@@ -0,0 +1,65 @@
+package co.omisego.omisego.model.transaction
+
+/*
+ * OmiseGO
+ *
+ * Created by Phuchit Sirimongkolsathien on 20/3/2018 AD.
+ * Copyright © 2017-2018 OmiseGO. All rights reserved.
+ */
+
+import co.omisego.omisego.model.pagination.Paginable
+import co.omisego.omisego.model.pagination.SortDirection
+
+/* Represent a structure used to query a list of transactions */
+data class TransactionListParams(
+        /**
+         * A page number
+         */
+        val page: Int,
+
+        /**
+         * A number of results per page.
+         */
+        val perPage: Int,
+
+        /**
+         * The sorting field
+         *
+         * The available values are
+         * - [Paginable.Transaction.SortableFields.ID]
+         * - [Paginable.Transaction.SortableFields.STATUS]
+         * - [Paginable.Transaction.SortableFields.FROM]
+         * - [Paginable.Transaction.SortableFields.CREATE_AT]
+         * - [Paginable.Transaction.SortableFields.UPDATED_AT]
+         */
+        val sortBy: Paginable.Transaction.SortableFields,
+
+        /**
+         * The desired sort direction
+         *
+         * The available values are
+         * - [SortDirection.ASCENDING]
+         * - [SortDirection.DESCENDING]
+         */
+        val sortDirection: SortDirection,
+
+        /**
+         * A term to search for in all of the searchable fields.
+         * See more at [Paginable.Transaction.SearchableFields]
+         *
+         * Note: Conflict with searchTerms, only use one of them.
+         */
+        val searchTerm: String? = null,
```

Comment:
Trully feels weird to have two params, one being `searchTerm` and the other `searchTerms` especially when they're exclusive.
Two ideas:

- Add 2 factory methods, one for each version, and make the constructor `internal`
- Find a way to have only one param 😂 (maybe with factory methods that'll wrap the single value in the map?)

Reply:
`search_term` is used to search in all the searchable fields while `search_terms` allows a map to be passed (`address: "123"`), therefore only searching in the specified field. Not sure what's the simplest way to represent that in the SDK 😅

Reply:
I guess the best solution would be 2 factory method, one for each case, and hide the constructor 🤔

Reply:
So it won't be `data class` anymore? We need to have a primary constructor for `data class` and even we mark parameters with `private`, they still can be initialized.

Reply:
I think for implementing that solution, the `TransactionListParam` may look like this

```kotlin
class TransactionListParams {
    /**
     * A page number
     */
    private var page: Int = 1

    /**
     * A number of results per page.
     */
    private var perPage: Int = 10

    /**
     * The sorting field
     *
     * The available values are
     * - [Paginable.Transaction.SortableFields.ID]
     * - [Paginable.Transaction.SortableFields.STATUS]
     * - [Paginable.Transaction.SortableFields.FROM]
     * - [Paginable.Transaction.SortableFields.CREATED_AT]
     * - [Paginable.Transaction.SortableFields.UPDATED_AT]
     */
    private var sortBy: Paginable.Transaction.SortableFields = Paginable.Transaction.SortableFields.CREATED_AT

    /**
     * The desired sort direction
     *
     * The available values are
     * - [SortDirection.ASCENDING]
     * - [SortDirection.DESCENDING]
     */
    private var sortDirection: SortDirection = SortDirection.DESCENDING

    /**
     * A term to search for in all of the searchable fields.
     * See more at [Paginable.Transaction.SearchableFields]
     *
     * Note: Conflict with searchTerms, only use one of them.
     */
    private var searchTerm: String? = null

    /**
     * A key-value map to search with the available fields
     * See more at [Paginable.Transaction.SearchableFields]
     *
     */
    private var searchTerms: Map<Paginable.Transaction.SearchableFields, Any>? = null

    /**
     * An optional address that belongs to the current user (primary address by default)
     */
    private var address: String? = null

    companion object {
        fun create(
                page: Int = 1,
                perPage: Int = 10,
                sortBy: Paginable.Transaction.SortableFields = Paginable.Transaction.SortableFields.CREATED_AT,
                sortDirection: SortDirection = SortDirection.DESCENDING,
                searchTerm: String? = null,
                address: String? = null
        ) = TransactionListParams().apply {
            this.page = page
            this.perPage = perPage
            this.sortBy = sortBy
            this.sortDirection = sortDirection
            this.searchTerm = searchTerm
            this.address = address
        }

        fun create(
                page: Int = 1,
                perPage: Int = 10,
                sortBy: Paginable.Transaction.SortableFields = Paginable.Transaction.SortableFields.CREATED_AT,
                sortDirection: SortDirection = SortDirection.DESCENDING,
                searchTerms: Map<Paginable.Transaction.SearchableFields, Any>? = null,
                address: String? = null
        ) = TransactionListParams().apply {
            this.page = page
            this.perPage = perPage
            this.sortBy = sortBy
            this.sortDirection = sortDirection
            this.searchTerms = searchTerms
            this.address = address
        }
    }
}
```

Reply:
It can be a data class :)

```kotlin
data class TransactionListParams private constructor(
   
    // Keep param public, but should probably be vals instead of vars?

) {
    companion object {
        fun create(
                page: Int = 1,
                perPage: Int = 10,
                sortBy: Paginable.Transaction.SortableFields = Paginable.Transaction.SortableFields.CREATED_AT,
                sortDirection: SortDirection = SortDirection.DESCENDING,
                searchTerm: String? = null,
                address: String? = null
        ) = TransactionListParams(
            page = page
            perPage = perPage
            sortBy = sortBy
            sortDirection = sortDirection
            searchTerm = searchTerm
            address = address
        }

        fun create(
                page: Int = 1,
                perPage: Int = 10,
                sortBy: Paginable.Transaction.SortableFields = Paginable.Transaction.SortableFields.CREATED_AT,
                sortDirection: SortDirection = SortDirection.DESCENDING,
                searchTerms: Map<Paginable.Transaction.SearchableFields, Any>? = null,
                address: String? = null
        ) = TransactionListParams(
            page = page
            perPage = perPage
            sortBy = sortBy
            sortDirection = sortDirection
            searchTerms = searchTerms
            address = address
        }
    }
}
```

Reply:
Only inconvenience of this is to have to declare default value twice, one for each factory method

Reply:
I think it's ok :)

Reply:
I think it should be fine 😀

Reply:
Changed it to data class